params/json.libevm.go

```diff
@@ -19,98 +19,129 @@ package params
+// UnmarshalJSON implements the [json.Unmarshaler] interface and JSON decodes
+// `data` according to the following:
+//   - extra is not registered:
+//     `data` is decoded into `c` and the extra is ignored.
+//   - extra is registered and the registered reuseJSONRoot field is false:
+//     `data` is decoded into `c` and the "extra" JSON field in `data` is decoded into `c.extra`.
+//   - extra is registered and the registered reuseJSONRoot field is true:
+//     `data` is decoded into `c` and `data` is decoded into `c.extra`.
+func (c *ChainConfig) UnmarshalJSON(data []byte) (err error) {
+	if !registeredExtras.Registered() {
+		return json.Unmarshal(data, (*chainConfigWithoutMethods)(c))
+	}
+	ec := registeredExtras.Get()
+	c.extra = ec.newChainConfig()
+	return UnmarshalChainConfigJSON(data, c, c.extra, ec.reuseJSONRoot)
 }
 
-// UnmarshalJSON implements the [json.Unmarshaler] interface.
-func (c *ChainConfig) UnmarshalJSON(data []byte) error {
-	switch reg := registeredExtras; {
-	case reg.Registered() && !reg.Get().reuseJSONRoot:
-		return c.unmarshalJSONWithExtra(data)
+// UnmarshalChainConfigJSON JSON decodes `data` according to the following.
+//   - `reuseJSONRoot` is false:
+//     `data` is decoded into `config` and the "extra" JSON field in `data` is decoded into `extra`.
+//   - `reuseJSONRoot` is true:
+//     `data` is decoded into `config` and `data` is decoded into `extra`.
+func UnmarshalChainConfigJSON[T any](data []byte, config *ChainConfig, extra *T, reuseJSONRoot bool) (err error) {
+	if extra == nil {
+		return fmt.Errorf("extra pointer argument is nil")
+	}
 
-	case reg.Registered() && reg.Get().reuseJSONRoot: // although the latter is redundant, it's clearer
-		c.extra = reg.Get().newChainConfig()
-		if err := json.Unmarshal(data, c.extra); err != nil {
-			c.extra = nil
-			return err
+	if reuseJSONRoot {
+		err = json.Unmarshal(data, (*chainConfigWithoutMethods)(config))
+		if err != nil {
+			return fmt.Errorf("decoding root chain config: %s", err)
 		}
-		fallthrough // Important! We've only unmarshalled the extra field.
-	default: // reg == nil
-		return json.Unmarshal(data, (*chainConfigWithoutMethods)(c))
+		err = json.Unmarshal(data, extra)
+		if err != nil {
+			return fmt.Errorf("decoding extra config to %T: %s", extra, err)
+		}
+		return nil
 	}
-}
 
-// unmarshalJSONWithExtra unmarshals JSON under the assumption that the
-// registered [Extras] payload is in the JSON "extra" key. All other
-// unmarshalling is performed as if no [Extras] were registered.
-func (c *ChainConfig) unmarshalJSONWithExtra(data []byte) error {
-	cc := &chainConfigWithExportedExtra{
-		chainConfigWithoutMethods: (*chainConfigWithoutMethods)(c),
-		Extra:                     registeredExtras.Get().newChainConfig(),
+	combined := struct {
+		*chainConfigWithoutMethods
+		Extra *T `json:"extra"`
+	}{
+		chainConfigWithoutMethods: (*chainConfigWithoutMethods)(config),
+		Extra:                     extra,
```

Comment:
```suggestion
		(*chainConfigWithoutMethods)(config),
		extra,
```
The direct coupling to the type definition and it not being a slice (i.e. this doesn't apply to table-drive tests IMO) makes the field names unnecessary and adds noise.

See [Concision](https://google.github.io/styleguide/go/guide#concision) style guide